cdc/sorter/leveldb/leveldb.go

```diff
@@ -0,0 +1,214 @@
+	wb       *leveldb.Batch
+	wbSize   int
+	iterSema *semaphore.Weighted
+	closedWg *sync.WaitGroup
+
+	metricWriteDuration prometheus.Observer
+	metricWriteBytes    prometheus.Observer
+}
+
+var _ actor.Actor = (*LevelActor)(nil)
+
+// NewLevelDBActor returns a leveldb actor.
+func NewLevelDBActor(
+	ctx context.Context, id int, db *leveldb.DB, cfg *config.SorterConfig,
+	wg *sync.WaitGroup, captureAddr string,
+) (*LevelActor, actor.Mailbox, error) {
+	idTag := fmt.Sprint(id)
+	// Write batch size should be larger than block size to save CPU.
+	wbSize := cfg.LevelDB.BlockSize * 16
+	wb := leveldb.MakeBatch(wbSize)
+	// IterCount limits the total number of opened iterators to release leveldb
+	// resources (memtables and SST files) in time.
+	iterSema := semaphore.NewWeighted(int64(cfg.LevelDB.Concurrency))
+	mb := actor.NewMailbox(actor.ID(id), cfg.LevelDB.Concurrency)
+	wg.Add(1)
+	return &LevelActor{
+		id:       actor.ID(id),
+		db:       db,
+		wb:       wb,
+		iterSema: iterSema,
+		wbSize:   wbSize,
+		closedWg: wg,
+
+		metricWriteDuration: sorterWriteDurationHistogram.WithLabelValues(captureAddr, idTag),
+		metricWriteBytes:    sorterWriteBytesHistogram.WithLabelValues(captureAddr, idTag),
+	}, mb, nil
+}
+
+func (ldb *LevelActor) close(err error) {
+	log.Info("leveldb actor quit", zap.Uint64("ID", uint64(ldb.id)), zap.Error(err))
+	ldb.closedWg.Done()
+}
+
+func (ldb *LevelActor) maybeWrite(force bool) error {
+	bytes := len(ldb.wb.Dump())
+	if bytes >= ldb.wbSize || (force && bytes != 0) {
+		startTime := time.Now()
+		err := ldb.db.Write(ldb.wb, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		ldb.metricWriteDuration.Observe(time.Since(startTime).Seconds())
+		ldb.metricWriteBytes.Observe(float64(bytes))
+
+		// Reset write batch or reclaim memory if it grows too large.
+		if cap(ldb.wb.Dump()) <= 2*ldb.wbSize {
+			ldb.wb.Reset()
+		} else {
+			ldb.wb = leveldb.MakeBatch(ldb.wbSize)
+		}
+	}
+	return nil
+}
+
+// Poll implements actor.Actor.
+// It handles tasks by writing kv, deleting kv and taking iterators.
+func (ldb *LevelActor) Poll(ctx context.Context, tasks []actormsg.Message) bool {
+	select {
+	case <-ctx.Done():
+		ldb.close(ctx.Err())
+		return false
+	default:
+	}
+	var iterChs []chan message.LimitedIterator
+	var iterRanges []*util.Range
+	for i := range tasks {
+		var task message.Task
+		msg := tasks[i]
+		switch msg.Tp {
+		case actormsg.TypeSorterTask:
+			task = msg.SorterTask
+		case actormsg.TypeStop:
+			ldb.close(nil)
+			return false
+		default:
+			log.Panic("unexpected message", zap.Any("message", msg))
+		}
+		events, needIter, iterCh := task.Events, task.NeedIter, task.IterCh
+
+		for k, v := range events {
+			if len(v) != 0 {
+				ldb.wb.Put([]byte(k), v)
+			} else {
+				// Delete the key if value is empty
+				ldb.wb.Delete([]byte(k))
+			}
+
+			forceWrite := false
+			if err := ldb.maybeWrite(forceWrite); err != nil {
+				log.Panic("leveldb error", zap.Error(err))
```

Comment:
why panic the actor goroutine here?

Reply:
In most cases, leveldb error means data corruption, we have no workaround, so panic.

Reply:
Can we distinguish whether it is data corruption error?

Reply:
No, I/O error may cause data corruption too, and it is different from `leveldb. ErrCorrupted`.